modyn/supervisor/internal/pipeline_executor/evaluation_executor.py

```diff
@@ -264,24 +294,28 @@ def worker_func(eval_req: EvalRequest) -> StageLog:
     # pylint: disable-next=too-many-locals
     def _single_batched_evaluation(
         self,
-        interval_start: int,
-        interval_end: Optional[int],
+        intervals: list[tuple[int, int | None]],
         model_id_to_eval: int,
         dataset_id: str,
-    ) -> tuple[Optional[str], dict]:
+    ) -> list[tuple[str | None, dict]]:
+        """Takes a list of intervals to be evaluated for a certain model and dataset.
+
+        Returns:
+            A list of tuples, where the first element is the failure reason (if any) and the second element is the
+            evaluation results. The order of the tuples corresponds to the order of the intervals.
+        """
         assert self.grpc.evaluator is not None, "Evaluator not initialized."
         assert self.pipeline.evaluation
         logger.info(
-            f"Evaluation Starts for model {model_id_to_eval} on split {interval_start} "
-            f"to {interval_end} of dataset {dataset_id}."
+            f"Evaluation Starts for model {model_id_to_eval} and dataset {dataset_id} on intervals:  {intervals}."
         )
         dataset_config = next((d for d in self.pipeline.evaluation.datasets if d.dataset_id == dataset_id))
 
         request = GRPCHandler.prepare_evaluation_request(
             dataset_config.model_dump(by_alias=True),
             model_id_to_eval,
             self.pipeline.evaluation.device,
-            intervals=[(interval_start, interval_end)],
+            intervals=cast(list[tuple[int | None, int | None]], intervals),
```

Comment:
why is this cast necessary? isn't this the input type of `intervals`?

Reply:
prepare_evaluation_request allows the start to be nullable, here it's nonnullable. And mypy doesn't allow narrowing those types in a nested setting.